### DISCUSSION TITLE
Saving library list As JSON file for personal catalog

### DISCUSSION THREAD

**Discussion by Sailfish** [audiamus/BookLibConnect#10] [Ideas]
After the users Library List is displayed but before running the conversion process, is there a way to allow the saving of the Library list as a single .json or .csv file? In viewing the temp files in AppData folder, it appears that you retrieve a large amount of data for each title, including such items as user reviews. However, limiting the fields to just what is displayed on the Book Lib Connect file list view (attached file) with the added features:

- Number of total titles in Library
- Limiting the catalog creation to just those titles selected
    This would allow quicker updating of the catalog (I have over 1300 titles)
- Only the user's ratings and review review text, if any
- The title synopsis text
- The local file link to the title image

![BookLibConnectLibraryList](https://user-images.githubusercontent.com/110050/151643656-4f5fa32d-f034-4b91-be5c-8dfc1db59b28.jpg)

**Comment by audiamus** [comment]
Exporting book meta data is on the wish list, but further down the line. Current priority are bug fixes, as the users report them, then integration into AAX Audio Converter.

The way meta data is retrieved and displayed works as follows:

- At the beginning of each new session, Book Lib Connect connects to your library on the Audible server, using the currently selected profile. It requests data for any new books that might have been added since your last purchase. At log level 3 (the default here), the reply is also stored in the tmp files, as LibraryResponse, you will probably have looked into it. The amount of information returned per book can be customized, but there is not much to gain there. 
  
- Most but not all of the response ends up in the local database. That's ordinary SQLite, a widely used single file relational database. My data model is a normalized one, with almost no redundancy but quite a few tables and several relationships, including multiplicities 1:1, 1:n and m:n. The author, for instance, gets stored in the Author table, not in the Book table. An author can have written more than one book and a book can have been written by more than one author. Genres or categories are special fun. Audible groups them in "ladders", consisting of "rungs" where each rung represents one specific category. Then we have single part and multi-part books. Every book or book part ("Component") has a "Conversion" associated with it, which will hold applied conversion parameters in the future, but mainly stores state and file destination at the moment. (The cover images are not stored in the database itself, only the file path to them.)

- What you see in the big list in the library window is again a subset of what is stored in the database. And it's flattened. To fit into reasonable space, only the first author is shown, the first narrator or the top genre. When you select a single book, more detailed info is shown in the three fields at the bottom. It's still all from the local database.

- If you select books for download, additional meta data is queried from the server. For multi-part books and if multi-part mode is on, size and duration for each part are called for, that's the ProductResponse in the tmp files. The new information goes to the database. Then, when you start the download for a book, the download location and the key is asked for. It is returned together with the detailed chapter data, again to be put into the database. That's the LicenseResponse. (The key is encrypted, so saving it as a file is pretty harmless.)

- The only information shown from these later queries to the server in either of the two lists are size and duration of parts and actual audio quality. (The books are always downloaded asking maximum quality but the results may vary.)

- What is slow when opening the library window the first time each session is the data retrieval from the local database, not from the server. As I said, that only queries new stuff, so usually minimum traffic. However, the query to the local database takes time, as the contents of several tables has to be read and converted from the relational into a proper object model that can be used as the data source for the library list.

**Reply by Sailfish** [reply]
As I have come to expect and always appreciate, your reply was thoughtful and detailed. I understand your priority to focus on Book Lib Connect bug fixes and AAX Audio Converter integration. I am also pleased to read of your desire to eventually provide a Library meta data export capability.

You are correct that I had examined the tmp/*.json files and noticed the "ladder and rungs" field redundancies and, quite frankly, was amazed at its magnitude.

I also noticed the SQLite db file and recall you mentioning using that database before. Is that db file (data/audiobooks.db) protected or can anyone open it if they have SQLite installed on their machine? The reason I ask is that I'm wary of Audible making changes to their site which may prevent AAX Audio Converter/Book Lib Connect from retrieving the meta data in the future and may want to make time to explore a slimmed version of the db library fields for an intermediate personal catalog until such time as may take you to provide a more polished version.

Respectfully,
Sailfish

**Reply by audiamus** [reply]
No, the database is not protected, there is no sensitive account data stored in it, only their alias(es), and all the book meta data, of course.. 

SQLite is just a single file, no database management system around it, so nothing that needs to be installed. All database interaction is performed by the applications themselves via driver libraries.  

You can inspect the database with any SQLlite tool, e.g. [SQLite Expert Personal](http://www.sqliteexpert.com/). Some tables are quite difficult to read, Primary/foreign relationships aren't that easy to trace. I would also urgently recommend not to modify any data.

**Comment by Sailfish** [comment]
A quick response to reassure you that I would not consider modifying any data which is part of either AAX Audio Converter or Book Lib Connect installations but, rather, export/save any derived data to my own file location(s).
